hyperon_das/tokenizers/dict_query_tokenizer.py

```diff
@@ -0,0 +1,190 @@
+            "and": [
+                DictQueryTokenizer.to_query_mapping[type(operand)](operand)
+                for operand in operator.operands
+            ],
+        },
+        NotOperator: lambda operator: {
+            "not": DictQueryTokenizer.to_query_mapping[type(operator.operand)](operator.operand),
+        },
+    }
+
+    @staticmethod
+    def tokenize(query: Query) -> str:
+        """
+        Convert a query dictionary into a tokenized string.
+
+        Args:
+            query (Query): The query dictionary to tokenize.
+
+        Returns:
+            str: A tokenized string representation of the query.
+
+        Raises:
+            ValueError: If the query cannot be tokenized.
+        """
+
+        def _tokenize(
+            _query: DictQueryTokenizer.Query, _parent: Link | None = None
+        ) -> Link | Node | Variable | OrOperator | AndOperator:
+            match _query:
+                case {"or": list()}:
+                    return OrOperator(operands=[_tokenize(o) for o in _query["or"]])
+                case {"and": list()}:
+                    return AndOperator(operands=[_tokenize(o) for o in _query["and"]])
+                case {"not": dict()}:
+                    return NotOperator(_tokenize(_query["not"]))
+                case {"atom_type": "link", "type": str(), "targets": list()}:
+                    link = Link(_query["type"])
+                    link.targets += [_tokenize(t, _parent=link) for t in _query["targets"]]
+                    return link
+                case {"atom_type": "node", "type": str(), "name": str()}:
+                    return Node(_query["type"], _query["name"])
+                case {"atom_type": "variable", "name": str()}:
+                    _parent.is_template = True
+                    return Variable(_query["name"])
+                case _:
+                    raise ValueError(f"Unsupported query: {_query}")
+
+        match query:
+            case {"or": list()} | {"and": list()} | {"not": dict()} | {
+                "atom_type": "link",
+                "type": str(),
+                "targets": list(),
+            }:
+                return TOKENS_DELIMITER.join(_tokenize(query).to_tokens())
+            case _:
+                raise ValueError(
+                    f"Unsupported query, it should start with a link or an operator: {query=}"
+                )
+
+    @staticmethod
+    def untokenize(query_tokens: str) -> Query:
+        """
+        Convert a tokenized string back into a query dictionary.
+
+        Args:
+            query_tokens (str): The tokenized string to untokenize.
+
+        Returns:
+            Query: A dictionary representation of the query.
+
+        Raises:
+            ValueError: If the tokens cannot be untokenized into a valid query.
+        """
+        cursor = 0
+        tokens = query_tokens.split()
+        if len(tokens) > 0 and tokens[0] in ("AND", "OR", "NOT", "LINK_TEMPLATE", "LINK"):
```

Comment:
Actually, "LINK" shouldn't be in this list.

Reply:
Removed.